src/entities/position/api/use-positions.ts

```diff
@@ -1,27 +1,55 @@
 import { DexService, ViewService } from '@penumbra-zone/protobuf';
 import { penumbra } from '@/shared/const/penumbra';
 import { connectionStore } from '@/shared/model/connection';
-import { useQuery } from '@tanstack/react-query';
+import { useInfiniteQuery } from '@tanstack/react-query';
 import {
   Position,
   PositionId,
-  PositionState_PositionStateEnum,
 } from '@penumbra-zone/protobuf/penumbra/core/component/dex/v1/dex_pb';
+import { AddressIndex } from '@penumbra-zone/protobuf/penumbra/core/keys/v1/keys_pb';
 import { bech32mPositionId } from '@penumbra-zone/bech32m/plpid';
 import { queryClient } from '@/shared/const/queryClient';
-import { AddressIndex } from '@penumbra-zone/protobuf/penumbra/core/keys/v1/keys_pb';
+
+const BASE_LIMIT = 20;
+const BASE_PAGE = 0;
+
+/** Helper function that allows breaking early when using stream responses */
+async function* limitAsync<T>(
+  iterable: AsyncIterable<T>,
+  limit: number,
+  offset: number,
+): AsyncIterable<T> {
```

Comment:
this function replaces basic `Array.fromAsync` and breaks early if the limit is reached, thus reducing calculations